rclpy/test/test_talker_listener.py

```diff
@@ -64,5 +71,52 @@ def test_talker_listener():
 
     async_launcher.terminate()
     async_launcher.join()
+    rclpy.shutdown()
+    assert len(received_messages) > 0, 'Should have received a message from talker'
 
-    assert len(received_messages) > 0, "Should have received a message from talker"
+
+def test_talker_listener_nested():
+    launch_desc = LaunchDescriptor()
+    launch_desc.add_process(
+        cmd=[sys.executable, os.path.join(this_dir, 'talker_nested.py')],
+        name='test_talker_listener_nested__talker'
+    )
+    launcher = DefaultLauncher()
+    launcher.add_launch_descriptor(launch_desc)
+    async_launcher = AsynchronousLauncher(launcher)
+    async_launcher.start()
+
+    import rclpy
+
+    rclpy.init([])
+
+    from rclpy.qos import qos_profile_default
+
+    from std_msgs.msg import Header
+    from geometry_msgs.msg import TransformStamped
+
+    assert Header.__class__._TYPE_SUPPORT is not None
+    assert TransformStamped.__class__._TYPE_SUPPORT is not None
+
+    node = rclpy.create_node('listener_nested')
+
+    received_messages = []
+
+    chatter_callback = functools.partial(
+        listener_nested_callback, received_messages=received_messages)
+    # TODO(wjwwood): should the subscription object hang around?
+
+    node.create_subscription(
+        TransformStamped,
+        'chatter_nested', chatter_callback,
+        qos_profile_default)
+
+    spin_count = 0
+    while len(received_messages) == 0 and spin_count < 10 and launcher.is_launch_running():
+        rclpy.spin_once(node)  # wait 1 second or until something has been done.
+        spin_count += 1
+
+    async_launcher.terminate()
+    async_launcher.join()
+    rclpy.shutdown()
+    assert len(received_messages) > 0, 'Should have received a nested_message from talker'
```

Comment:
This functions seems to contain a lot of copy-n-pasted code. Please try to refactor it to reuse common code blocks.